rdpy/testlib/test_client.py

```diff
@@ -0,0 +1,137 @@
+from rdpy.protocol.rdp import rdp
+import rdpy.core.log as log
+from threading import Event
+import socket
+
+
+class RDPTestClient(rdp.RDPClientObserver):
+    """
+    @summary: Connection Object that receives session events from the twisted
+    stack and the remote server
+    """
+
+    def __init__(self, controller,
+                 width, height,
+                 startedEvent, stoppedEvent):
+        rdp.RDPClientObserver.__init__(self, controller)
+        controller.setScreen(width, height)
+
+        self.hasInitialised = False
+        self.hasSession = False
+        self._startedEvent = startedEvent
+        self._stoppedEvent = stoppedEvent
+
+    def onReady(self):
+        """
+        @summary: Called when stack is ready
+        """
+        self.hasInitialised = True
+
+    def onSessionReady(self):
+        """
+        @summary: Windows Session Reported Ready
+        """
+        self.hasSession = True
+        self._startedEvent.set()
+
+    def onClose(self):
+        """
+        @summary: Called when the connection parts are closed and sets the
+        stoppedEvent
+        """
+        self.hasSession = False
+        self.hasInitialised = False
+        self._stoppedEvent.set()
+
+    def onUpdate(self, destLeft, destTop, destRight, destBottom, width, height, bitsPerPixel, isCompress, data):
+        """
+        @summary: callback use when bitmap is received
+        """
+        pass
+
+    @property
+    def loggedIn(self):
+        """
+        Reports that we have a session with the target server
+        """
+        return self.hasInitialised and self.hasSession
+
+    def close(self):
+        """
+        @summary close this connection
+        """
+        self._controller.close()
+
+
+class RDPTestClientFactory(rdp.ClientFactory):
+    """
+    @summary: Builds connection to an RDP Server.
+    """
+
+    def __init__(self, username, password, startedEvent):
+        self._username = username
+        self._passwod = password
+
+        self.stopped = False
+        self.reason = None
+        self._startedEvent = startedEvent
+        self._stoppedEvent = Event()
```

Comment:
Does `rdpy` itself use threading events? Given that it's twisted, I'd have said that we should be using `Deferred`s instead, even if for nominally the same pattern.